WIP: Add `ITK_GAUSSIAN` generation option

tbirdso opened the pull request:
Adds Gaussian blurring procedure for multiscale pyramid generation roughly following the steps outlined in `itk::MultiResolutionPyramidImageFilter`:
1. Run Gaussian blurring on input image
2. Resample result for given scale factors

Currently only testing on 2D input. @thewtex Could you please weigh in on the following issues that I am seeing?
1. Baseline comparison fails because there is a difference in output coordinates. I assume this means that there is an issue in resampling. Do you have thoughts on what correction is needed here? Maybe the origin needs to be shifted slightly based on scale factors?
```py
E           AssertionError: Left and right Dataset objects are not equal
E
E           Differing coordinates:
E           L * y        (y) float64 0.5 2.5 4.5 6.5 8.5 ... 246.5 248.5 250.5 252.5 254.5
E           R * y        (y) float64 0.0 2.0 4.0 6.0 8.0 ... 246.0 248.0 250.0 252.0 254.0
E           L * x        (x) float64 0.5 2.5 4.5 6.5 8.5 ... 246.5 248.5 250.5 252.5 254.5
E           R * x        (x) float64 0.0 2.0 4.0 6.0 8.0 ... 246.0 248.0 250.0 252.0 254.0
E           Differing data variables:
E           L   cthead1  (y, x) uint8 0 0 0 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0 0 0
E           R   cthead1  (y, x) uint8 dask.array<chunksize=(128, 128), meta=np.ndarray>

test\test_spatial_image_multiscale.py:50: AssertionError
```

Note that `itk.ResampleImageFilter` expects us to request output metadata, so I have calculated output size/spacing/origin explicitly rather than using `filter.UpdateOutputInformation` as done in the `Methods.ITK_BIN_SHRINK` procedure.

2. I am unclear on how to write out a single scale image result to disk in a way that can be visualized with 3D Slicer or a similar application. Could you please clarify or provide a code snippet detailing how to write out output results for visual evaluation?

This is a first stab at introducing gaussian blurring and does not currently make use of FFTs or VkFFT acceleration. These can be added in an incremental update, possibly using [VkDiscreteGaussianImageFilter](https://github.com/InsightSoftwareConsortium/ITKVkFFTBackend/pull/51).

thewtex commented on spatial_image_multiscale.py:
```diff
@@ -313,6 +314,140 @@ def align_chunks(current_input, dim_factors):
+
+        def compute_sigma(input_spacings, shrink_factors) -> list:
+            assert len(input_spacings) == len(shrink_factors)
+            output_spacings = [input_spacing * shrink for input_spacing, shrink in zip(input_spacings, shrink_factors)]
```
Can we put a reference to the logic here in a comment?

tbirdso replied:
Added

thewtex commented on spatial_image_multiscale.py:
```diff
@@ -313,6 +314,140 @@ def align_chunks(current_input, dim_factors):
+            dim_factors = dim_scale_factors(scale_factor)
+            current_input = align_chunks(current_input, dim_factors)
+
+            image_dims: Tuple[str, str, str] = ("x", "y", "z", "t")
```
4 `str`'s

tbirdso replied:
Fixed

thewtex commented:
@tbirdso looking great!

>  I assume this means that there is an issue in resampling

For resampling, we could use this strategy:

1. Filter with discrete gaussian
2. Use BinShrink UpdateOutputInformation to determine the sampling grid
3. Use the results of 1. and 2. as the input image and the definition for the sampling grid in a resample image filter

> I am unclear on how to write out a single scale image result to disk in a way that can be visualized with 3D Slicer or a similar application. Could you please clarify or provide a code snippet detailing how to write out output results for visual evaluation?

```
image_da = multiscale['scale1'].ds.cthead1
itk_image = itk.image_from_xarray(image_da)
itk.imwrite(itk_image, 'itk_image.nrrd')
````

thewtex commented on spatial_image_multiscale.py:
```diff
@@ -313,6 +314,140 @@ def align_chunks(current_input, dim_factors):
+                              for input_spacing, shrink_factor in zip(itk.spacing(block_0_input), shrink_factors)]
+            dtype = block_0.dtype
+
+            blurred_array = map_overlap(
```
Perhaps we could put all three steps in this call in a Python function?

tbirdso commented:
@thewtex Thanks for the help, interesting that we are getting different output metadata from `BinShrinkImageFilter` as opposed to output following the `MultiResolutionPyramidImageFilter` procedure. I may need to dig into `BinShrinkImageFilter` a bit more to understand what is happening.

I've updated with the suggestions above, while I am getting output for each level I am seeing tenuous results. I am trying to pin down where exactly my logic is going wrong, I have added an optional 2D isotropic image printout from testing for visual comparison of outputs. I have a few questions about previous output baselines and the draft procedure
1. I am a bit confused on how we are defining "scale factors" here. For both existing procedures the baseline image has a spacing of 1.0mm, an isotropic scale of 2 has a spacing of 2.0mm, but then an isotropic scale of 4 has a spacing of 6.0mm. The image extent is kept roughly the same at each level with sizes of 256, 128, and then 42. I would expect that spacing scales with the scale factor (i.e. 1mm, 2mm, 4mm) and the image size follows suit (256, 128, 32). Am I misunderstanding the procedure at hand?
2. Under the proposed Gaussian+downsampling implementation, for 2D image input with isotropic scale factors of [2,4], the first scale image appears roughly correct with a slight translation, but the second scale image is greatly downsized and occupies only the top left corner of the 32x32 image. Both describe the same image extent, so it doesn't look like `ResampleImageFilter` is simply being given the wrong output. Do you have any thoughts as to where the issue may lie?

Scale1 (2.0mm spacing):

![gaussian1](https://user-images.githubusercontent.com/40648863/167216850-25863cf9-33f4-422d-b82a-80e1be74b477.png)

Scale2 (4.0mm spacing):

![gaussian2](https://user-images.githubusercontent.com/40648863/167216834-9d9a06dd-b2b8-4aaf-923b-6962d8a97e2e.png)

tbirdso commented:
Aha, I've identified two issues due to applying ITK filters to chunks that are passed in in an absence of metadata:
- The origin for resampling must be shifted by the kernel radius to compensate for padding
- Output spacing must be set as a ratio of input to output spacing per shrink factors rather than passing in the actual output spacing, because chunks are assumed to have isotropic spacing of 1 when converted from xarray

Will prepare a patch. @thewtex Do you have advice on how to add new baseline image data via IPFS?

thewtex commented:
@tbirdso fantastic!

Please take a look at testing data docs here: https://github.com/spatial-image/multiscale-spatial-image/pull/36

thewtex commented:
> 256, 128, and then 42

Yes, this should be 256, 128, 64, with spacing / scale set accordingly!

tbirdso commented:
Code is largely functional, two lingering issues:
1. `itk-vkfft` packages are building for Linux 3.10, required for kernel radius computation and FFT GPU speedup
2. Output result is as expected, but uncertain that Dask processing is executing over chunks, printouts seem to show that entire image is passed to `map_overlap`. May require further investigation and testing with larger image.

tbirdso commented:
Looks like something strange is going on in the ITKVkFFTBackend Windows modules. I was able to reproduce the issue in [https://github.com/InsightSoftwareConsortium/ITKVkFFTBackend/runs/6357790869?check_suite_focus=true](https://github.com/InsightSoftwareConsortium/ITKVkFFTBackend/runs/6357790869?check_suite_focus=true):
```
Loading ITKFiniteDifference... done
Loading ITKDisplacementField... done
Loading ITKRegistrationCommon... done
Loading VkFFTBackend... Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\lazy.py", line 138, in __getattribute__
    base.itk_load_swig_module(module, namespace)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 96, in itk_load_swig_module
    itk_load_swig_module(dep, namespace)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 96, in itk_load_swig_module
    itk_load_swig_module(dep, namespace)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 241, in itk_load_swig_module
    load_module_needed_factories(name)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 385, in load_module_needed_factories
    load_factories(needed_factories[name])
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 365, in load_factories
    itk_load_swig_module(module_name,namespace)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 132, in itk_load_swig_module
    l_module = loader.load(swig_module_name)
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\base.py", line 291, in load
    l_spec.loader.exec_module(l_module)  # pytype: disable=attribute-error
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\hostedtoolcache\windows\Python\3.10.4\x64\lib\site-packages\itk\support\..\VkFFTBackendPython.py", line 13, in <module>
    from . import _VkFFTBackendPython
ImportError: DLL load failed while importing _VkFFTBackendPython: The specified module could not be found.
Error: Process completed with exit code 1.
```

The issue only occurs for `windows-2019` runners, Ubuntu and macOS observed to pass all pytests without failure. I am not able to reproduce the issue with a clean Python 3.8 or 3.10 environment on my local Windows machine. Tmate remote debugging appears to be bogged down today and is not able to get a connection for me to investigate the failure. Will continue to investigate to the best of my ability using available resources.

thewtex commented on multiscale_spatial_image.py:
```diff
@@ -313,6 +314,168 @@ def align_chunks(current_input, dim_factors):
+            sigma_values = compute_sigma(itk.spacing(image), shrink_factors)
+            # Constrain kernel width to be at most the size of one chunk
+            max_kernel_width = min(DEFAULT_MAX_KERNEL_WIDTH, *itk.size(image))
+            filt = itk.VkDiscreteGaussianImageFilter.New(
```
Something like:

```
if hasattr(itk, 'VkDiscreteGaussianImageFilter'):
    Filt =  itk.VkDiscreteGaussianImageFilter
else:
    Filt = itk.DiscreteGaussianImageFilter

filt = Filt.New(*args)
```

thewtex commented on multiscale_spatial_image.py:
```diff
@@ -313,6 +314,168 @@ def align_chunks(current_input, dim_factors):
+            
+            # Construct pipeline
+            sigma_values = compute_sigma(input_spacing, shrink_factors)
+            smoothing_filter = itk.VkDiscreteGaussianImageFilter.New(image, 
```
Vk if available

thewtex commented on test/test_multiscale_spatial_image.py:
```diff
@@ -59,9 +59,17 @@ def test_base_scale(input_images):
+    
+def store_new_image(multiscale_image, dataset_name, baseline_name):
+    '''Helper method for writing output results to disk
+       for later upload as test baseline'''
```
👍

thewtex reviewed (commented):
Love it!

A few comments inline

thewtex commented:
> Looks like something strange is going on in the ITKVkFFTBackend Windows modules.

Can we use the CPU version by default (see review comments), then the GPU version if available?

I do not think the Windows GitHub Actions runners have a GPU or OpenCL support.

We could add a separate GitHub Actions test configuration for GPU testing as done with the ITKVtkFFTBackend module.

tbirdso commented on multiscale_spatial_image.py:
```diff
@@ -313,6 +315,166 @@ def align_chunks(current_input, dim_factors):
+    elif method is Methods.ITK_GAUSSIAN:
+        import math
+        import itk
+        assert 'VkFFTBackend' in dir(itk)  # Verify itk-vkfft installed
```
The underlying `DiscreteGaussianImageFilter` does not publicly expose useful methods such as `GetKernelRadius` that I have used from `VkDiscreteGaussianImageFilter`. But yes, I can add a workaround to compute these values explicitly for multiscale generation.

tbirdso commented:
@thewtex 

> Can we use the CPU version by default (see review comments), then the GPU version if available?

Yes, this is feasible though has a few extra steps to adjust the current implementation beyond selecting between filters. I will update accordingly.

> I do not think the Windows GitHub Actions runners have a GPU or OpenCL support.

I agree that Windows runners have no GPU support but disagree that they have no OpenCL support. If that were the case then our ITKVkFFTBackend Windows CI would always fail (see [https://github.com/InsightSoftwareConsortium/ITKVkFFTBackend/runs/6353049662?check_suite_focus=true](https://github.com/InsightSoftwareConsortium/ITKVkFFTBackend/runs/6353049662?check_suite_focus=true)). 

Either way we \*should\* be packaging the OpenCL-ICD-Loader with `itk-vkfft` such that build/load will succeed and then actually running will fail if OpenCL is not available on the underlying platform. It's possible that the failure is a result of the OpenCL-ICD-Loader not being properly packaged and distributed with Windows wheels. 

> We could add a separate GitHub Actions test configuration for GPU testing as done with the ITKVtkFFTBackend module.

I agree that this is a good idea for down the road.

tbirdso commented:
@thewtex Seeing poor performance fetching IPFS data for tests (30+ minutes to pin). Do you have suggestions on how to alleviate/improve slowdowns?

thewtex commented:
> @thewtex Seeing poor performance fetching IPFS data for tests (30+ minutes to pin). Do you have suggestions on how to alleviate/improve slowdowns?

We can also pin on pinata, then connect to these peers.

thewtex commented on multiscale_spatial_image.py:
```diff
@@ -313,6 +314,182 @@ def align_chunks(current_input, dim_factors):
+            '''Blur and then downsample a given image chunk'''
+
+            # xarray chunk does not have metadata attached, input values are ITK defaults
+            image = itk.image_from_array(xarray_data)
```
`image_view_from_array` to avoid memory allocation / copy

thewtex reviewed (commented):
Looks great! One performance issue noted inline.

thewtex reviewed (approved):
:small_airplane:
